README.md
## Angular + Firebase + Nodejs + ELK + Docker + Docker-Compose + RSS + Travis
[![Code Climate](https://codeclimate.com/github/sloppylopez/docker_elk_stack/badges/gpa.svg)](https://codeclimate.com/github/sloppylopez/docker_elk_stack) Flawless Victory, Fatality!

[![Travis](https://travis-ci.org/sloppylopez/docker_elk_stack.svg)](https://travis-ci.org/sloppylopez/docker_elk_stack)

[![Test Coverage](https://codeclimate.com/github/sloppylopez/docker_elk_stack/badges/coverage.svg)](https://codeclimate.com/github/sloppylopez/docker_elk_stack/coverage) Fair enough :_(

<a href="https://github.com/sloppylopez/docker_elk_stack/releases/tag/v0.0.9-feel-like-a-sir">v0.0.9-feel-like-a-sir</a> Released! 

![alt tag](http://i3.kym-cdn.com/photos/images/original/000/161/140/feel-like-a-sir-template.jpg)

##FEATURES SO FAR:
1) CI and CD with Travis, Code climate and Firebase, plus code climate it's integrated with github issues
   so you can create a github issue when quality of code decreased after a PR in 1 click

2) Scaffolding of angular client done with Yeoman (default angular generator)

3) User management, including email verification fully integrated using angularfire

4) Plain default mild installation of all components, bower, npm, docker which
  means the whole project is mainstreamed at maximum, no hacky shit, no wheel
  reinventions (at least that i am aware of)
  
5) Run in almost all devices and browsers(IE8+ and all the rest) and it's responsive

6) Completely logicless form validation using new angular 1.3 directive ng-messages

This 2 lines make all the validation, including displaying the custom error messages for each case
which completely erases the boiler plate of form validation, this is the so called 'angular way'
>< input id="username" type="email" name="email" class="form-control" ng-model="user.email" required email/>

>< div ng-messages="loginForm.email.$error" ng-messages-include="views/errors.html" class="errors"></div>
   
##INTENDED FEATURES:   

1)Manage jenkins and docker cluster with rancher

##INTRO:
Front-End Angular, Back-End Nodejs, Docker running 3 servers with ELK stack orchestrated with Docker-compose

This project it's 3 virtual machines build with docker and docker-compose for the orchestration of the 3 services, Elastic search, Logstash and Kibana, also there is a front-end angular client, and a backend nodejs that includes a elastic search client to consume rss from any website and persist them in the EL in a bulk process, then you can visualize them in the angular frontend and do the tipical text search over the rss's

It can be used as a seed for a project with that structure or similar, it saves a lot of 'default work' to set everything together.

There are 2 main goals in this project

1)Create a 'seed' that can be reused everytime you need to build a web-app with similar characteristics
like in this case angular, node and elk

2)Use the finished version of this project to build something over it, like for example without
entering yet in many details, extract info from the rss news stored with Spark or a similar technology
IN REAL TIME

This are the 3 separate parts :

##angularclient
Implemented using Yeoman, mild default installation, it uses Karma for testing and Grunt for building
Includes angular-fire a implementation of firebase client for angular, we use it as a Baas (Backend
as service), this is much more than a database, you can deploy you project in production with 1
command, it has a nice web interface to maintain the back-end, including non-google-web-analytics,
user management, testing console, and many nice options that erase the boiler plate of 'deploying in
prod'/'managing a server'/'managing a backend' for your FRONT-END application, it simply works out of the box
the project has implemented USER EMAIL VALIDATION using firebase Baas magic and angularfire modules.
###*See how to run below

##nodeapi
Implemented using Nodejs, it gets rss from any source and bulks the data in Elastic Search using json format, this part will become a Nodejs module in the future, and I will use a RESTful api, but that part it's still to be defined, this is just a node server with Express 4 and a Elastic Search client with a Xml parser for retrieving, parsing and persisting in EL
###*Runs with docker-compose in the elk-stack

##dockers
This is the FUN part, based on this article (http://thepracticalsysadmin.com/running-elk-on-docker/)
It's 3 virtual machines one per each element of the stack, connected between them with minimum configuration, if you check inside each folder there is one the Dockerfile to create each
container, plus the instructions to download and configure,USING the rest of the files of each folder to have a complete running setup once all dependencies have been downloaded by Docker
###*Use the run scripts in the folder to run it, it works out of the box.

PULL AND PLAY!!!!

(_̅_̅_̅_̅(̅_̅_̅_̅_̅_̅_̅_̅_̅_̅_̅_̅_̅()ڪ

###Keep it stupid!, simple.

![alt tag](http://lh3.ggpht.com/ZpL4xc4OOMqiKxrLzJ3w33AOlu5ERX4ZbAsEWqSBQWeLYLF2hybEw11D5msBRLtGY1DoKYMI-A8VBx3yvwpNNUmr=s250)

### To build the whole stack and client:
I only give Linux support for the moment, in the future I will probably cover Windows as well.
I want you to understand that this project it's literally building 4 servers, each one configured with
1 different and complementary technology, which are visible between them with minimum configuration
which at the same time integrates with a dockerized built CD and CI using jenkins, as you can imagine
doing all this with only 10 steps it's really a time saver, but also I am saying this to make you understand
that i noticed when installing sometimes problems with components, specially with grunt, when doing the steps
when had to execute 'grunt serve' it gives an wierd error, and the project does not launch, this has happened
several times with the testing of the 100x100 out of the box working installation, however, after deleting
node_modules, .temp, dist and bower_components and reinstalling them again it simply started to work,
this is the only pitfall i can find in the instalation, just to make you aware. so if you see this try what i
told you and it will work

Prerequisites :

    > Linux
    > Python >=2.6
    > Nodejs >=0.12.0
    > Npm
    > Bower
    > Git
    > Docker >=1.6.0
    > Karma
    > Grunt
    > Docker-compose
    > Ruby

### To build:FIRST INSTALL!
    Notice that 1) it's intended for plain Ubuntu, if you have already java, npm and so on, 
    you should not use 1) because it can mess up your working environment, my suggestion to 
    try this is to use Virtual Box with a clean ubuntu install, If you met 
    pre-requisites it should work without 1) but i can't test so many scenarios so im not
    100% sure what would happen if you try to install it without 1)
    
    0) delete the 'env: global: secure:' section in .travis.yml 
    1) modify prerequisites_install.sh to add your credentials
    2) sudo sh prerequisites_install.sh
    3) sudo sh run_dockers.sh
       if you get an error saying docker daemon is not up just do 'docker -d' and do 2) again
    4) change angularclient/script/constants.js to match your own values, 
       basically after 2) you can discover the docker ip doing ifconfig and such.
    5) sudo npm install;bower install
    6) mkdir dist
       (to create the dist directory which we will use to deploy in firebase)
    7) firebase init
    8) grunt serve
    9) Load rss in elasticsearch with fresh rss news:
       http://localhost:8081/api/rssload
    10) go to https://travis-ci.org/<YOUR_GIT_USER>/<YOUR_GIT_PROJECT_NAME>/settings/env_vars 
        and create 3 env variables and add this env vars
         - CODECLIMATE_REPO_TOKEN=<YOUR_CODECLIMATE_REPO_TOKEN>
         - FIREBASE_USER=<YOUR_FIREBASE_USER>
         - FIREBASE_PASSWORD=<YOUR_FIREBASE_PASSWORD>
       
    *NOTES:
        0) It for removing my secure keys before you do 2) which
           will automatically generate them and add them to your .travis.yml
        2) removes the boiler plate of the dependencies
        3) builds and runs the backend (ELK Stack)
        4)5)8) it's for running the client, to visualize the data
           that came from nodeapi to elasticsearch, and from elastic 
           search to angular, if you can see the rss news in angular,
           it's complete correctly instaled
        6) creates a release inside 'dist' folder, needed for firebase
        7) initialize firebase (only first time, this is only 
           needed for deploying angularclient in firebase hosting)

###To deploy in production:
    This is a continuous delivery project, so there is no such 
    'deploy to prod' anymore, however for learning purposes here 
    is how you do it handcraft:
    
    Pre-requisites:
        A) Firebase account (free)
        B) 'firebase init' command executed in angularclient. he will 
           promp you questions, i recommend to choose the folder ./dist 
           as the folder firebase will deploy in prod since 'grunt build' 
           will put the minified version of our js and css there
        C) Change constants.js in angularclient to match your credentials
        
    To deploy:    
        1)grunt build
        2)firebase deploy

###After install:
        1) http://localhost:9000/#/    angularclient
        2) http://localhost:5601       kibana
        3) http://localhost:8081/api   nodeapi
        4) http://localhost:9200       elasticsearch
        5) http://localhost:8080       jenkins
    
###Next steps:(Changed priorities 25-may-2015)
    1)Start using rancherOS for cooking the images, 
    so i will change Kibana, Logstash and EL to start
    using rancherOS(Dedicated docker OS) instead of
    phusion image we are using right now, i think
    this is the way to go
    
    2)Start using rancher to manage jenkins slaves

    3)Finish test coverage
           
    4)Finish test e2e with protractor

    5)Deploy docker back-end ,i am not quite sure how
    much ram will need this to work, i will test calmly
    using virtual box but this process can take time,
    because i am going to try to deploy it for
    free, not sure if this can de done easily, or if will
    have to give up and pay :_(
    
    6)Merge with https://github.com/firebase/firereader
    this will eliminate all the boiler plate of managing and displaying
    nicely the feeds, because the real intention of this project is not
    displaying rss through a Elastic Search, but to analyze those rss
    to extract information using Spark or similar.
    
    7)I am studying to add a plug-in for Grunt for automatically 
    resizing images, so we can avoid the boiler plate of having 
    to manually resize images in our project using Photoshop or 
    similar on every image addition.
    
###Info sources:
    https://johnheroy.com/2014/10/17/continuous-firebase-deployment-with-travis.html
    http://adamwathan.me/2014/12/28/test-coverage-code-climate-travis-ci
    http://docs.travis-ci.com/user/encryption-keys/
    https://www.firebase.com/docs/hosting/command-line-tool.html
    https://www.npmjs.com/package/codeclimate-test-reporter
    http://docs.travis-ci.com/user/caching/
    http://docs.travis-ci.com/user/notifications/#Webhook-notification


changes.md
## 0.0.9-feel-like-a-sir TBR
    KNOWN BUGS:
        * Nasty issue that i don't like, i will think what to do with this
          and prepare travis to test ONLY angular, because it seems he it's not able to test
          the docker build, shame...
          https://github.com/travis-ci/travis-ci/issues/1196
          However we will fix this using Jenkins
        * Failed to download bootstrap-sass-official dependency in the CD
          process, but it is just a minor

    FEATURES ADDED:
        * 100% coverage integration test
        * Protractor e2e test

## 0.0.8-feel-like-a-sir DO NOT USE THIS VERSION! I've discovered docker won't run because of Jenkins
    KNOWN BUGS:
        * Nasty issue that i don't like, i will think what to do with this
          and prepare travis to test ONLY angular, because it seems he it's not able to test
          the docker build, shame...
          https://github.com/travis-ci/travis-ci/issues/1196
          However we will fix this using Jenkins
        * Failed to download bootstrap-sass-official dependency in the CD
          process, but it is just a minor
        * Failed to download bootstrap-sass-official dependency in the CD
          process, but it is just a minor

    FIXED BUGS:
        * Incorrect Firebase params in constants.js

## 0.0.6-feel-like-a-sir

    KNOWN BUGS:
        * Nasty issue that i don't like, i will think what to do with this
          and prepare travis to test ONLY angular, because it seems he it's not able to test
          the docker build, shame...
          https://github.com/travis-ci/travis-ci/issues/1196
          However we will fix this using Jenkins

    FEATURES ADDED:
        * Added full Travis CI and CD integration

## 0.0.5-feel-like-a-sir

    KNOWN BUGS:
        * Nasty issue that i don't like, i will think what to do with this
          and prepare travis to test ONLY angular, because it seems he it's not able to test
          the docker build, shame...
          https://github.com/travis-ci/travis-ci/issues/1196
          
    KNOWN BUGS:          
        * Fixed login bug in angular client, now you can log 
          from any device

## 0.0.4-feel-like-a-sir

    KNOWN BUGS:
        * Can't log in with mobile devices

    BUGS FIXED:
        * Cleaned and documented and running out of the box

## 0.0.3-feel-like-a-sir

    KNOWN BUGS:
        * Can't log in with mobile devices

    BUGS FIXED:
        * Added Karma to prerequisites_install.sh

## 0.0.2-feel-like-a-sir

    KNOWN BUGS:
        * Can't log in with mobile devices

    BUGS FIXED:
        * Working 100x100 out of the box following the instructions

## 0.0.1-feel-like-a-sir

    KNOWN BUGS:
        * Can't log in with mobile devices

    BUGS FIXED:
        * Wrong redirection after authWithPassword
        * Spinner was not fitting perfectly again...

## 0.0.1-alpha

    BUGS FIXED:
        * Fixed failure in docker build the stop of jdk distribution from docker repo
        * Spinner was not fitting perfectly.

    FEATURES:Build frameworks working:Docker, Grunt and Yeoman
             Test framework working:Karma combined with Mocha and Chai
             Angular-Firebase integration 50%, remains undone 'chat' and 'three-way data binding'
             Angular client working including:
                Full user authentication
                Rss filtering using the elastic search in the docker stack
                Kibana working
                Logstash working
                Elastic Search working
                Nodeapi persisting rss in the EL

    KNOW ISSUES:The docker elk stack it's not online yet... i will deploy in Heroku
                or AWS ASAP!!!! :)


angularclient/Gruntfile.js
// Generated on 2015-04-25 using
// generator-webapp 0.5.1
'use strict';

// # Globbing
// for performance reasons we're only matching one level down:
// 'test/spec/{,*/}*.js'
// If you want to recursively match all subfolders, use:
// 'test/spec/**/*.js'

module.exports = function (grunt) {

  // Time how long tasks take. Can help when optimizing build times
  require('time-grunt')(grunt);

  // Load grunt tasks automatically
  require('load-grunt-tasks')(grunt);

  // Configurable paths
  var config = {
    app: 'app',
    dist: 'dist'
  };

  // Define the configuration for all the tasks
  grunt.initConfig({

    // Project settings
    config: config,

    // Watches files for changes and runs tasks based on the changed files
    watch: {
      bower: {
        files: ['bower.json'],
        tasks: ['wiredep']
      },
      js: {
        files: ['<%= config.app %>/scripts/{,*/}*.js'],
        tasks: ['jshint'],
        options: {
          livereload: true
        }
      },
      jstest: {
        files: ['test/spec/{,*/}*.js'],
        tasks: ['test:watch']
      },
      gruntfile: {
        files: ['Gruntfile.js']
      },
      sass: {
        files: ['<%= config.app %>/styles/{,*/}*.{scss,sass}'],
        tasks: ['sass:server', 'autoprefixer']
      },
      styles: {
        files: ['<%= config.app %>/styles/{,*/}*.css'],
        tasks: ['newer:copy:styles', 'autoprefixer']
      },
      livereload: {
        options: {
          livereload: '<%= connect.options.livereload %>'
        },
        files: [
          '<%= config.app %>/{,*/}*.html',
          '.tmp/styles/{,*/}*.css',
          '<%= config.app %>/images/{,*/}*'
        ]
      }
    },

    // The actual grunt server settings
    connect: {
      options: {
        port: 9000,
        open: true,
        livereload: 35729,
        // Change this to '0.0.0.0' to access the server from outside
        hostname: 'localhost'
      },
      livereload: {
        options: {
          middleware: function(connect) {
            return [
              connect.static('.tmp'),
              connect().use('/bower_components', connect.static('./bower_components')),
              connect.static(config.app)
            ];
          }
        }
      },
      test: {
        options: {
          open: false,
          port: 9001,
          middleware: function(connect) {
            return [
              connect.static('.tmp'),
              connect.static('test'),
              connect().use('/bower_components', connect.static('./bower_components')),
              connect.static(config.app)
            ];
          }
        }
      },
      dist: {
        options: {
          base: '<%= config.dist %>',
          livereload: false
        }
      }
    },

    // Empties folders to start fresh
    clean: {
      dist: {
        files: [{
          dot: true,
          src: [
            '.tmp',
            '<%= config.dist %>/*',
            '!<%= config.dist %>/.git*'
          ]
        }]
      },
      server: '.tmp'
    },

    // Make sure code styles are up to par and there are no obvious mistakes
    jshint: {
      options: {
        jshintrc: '.jshintrc',
        reporter: require('jshint-stylish')
      },
      all: [
        'Gruntfile.js',
        '<%= config.app %>/scripts/{,*/}*.js',
        '!<%= config.app %>/scripts/vendor/*',
        'test/spec/{,*/}*.js'
      ]
    },

    // Compiles Sass to CSS and generates necessary files if requested
    sass: {
      options: {
        sourceMap: true,
        includePaths: ['bower_components']
        },
      dist: {
        files: [{
          expand: true,
          cwd: '<%= config.app %>/styles',
          src: ['*.{scss,sass}'],
          dest: '.tmp/styles',
          ext: '.css'
        }]
      },
      server: {
        files: [{
          expand: true,
          cwd: '<%= config.app %>/styles',
          src: ['*.{scss,sass}'],
          dest: '.tmp/styles',
          ext: '.css'
        }]
      }
    },

    // Add vendor prefixed styles
    autoprefixer: {
      options: {
        browsers: ['> 1%', 'last 2 versions', 'Firefox ESR', 'Opera 12.1']
      },
      dist: {
        files: [{
          expand: true,
          cwd: '.tmp/styles/',
          src: '{,*/}*.css',
          dest: '.tmp/styles/'
        }]
      }
    },

    // Automatically inject Bower components into the HTML file
    wiredep: {
      app: {
        ignorePath: /^\/|\.\.\//,
        src: ['<%= config.app %>/index.html'],
        exclude: ['bower_components/bootstrap-sass-official/assets/javascripts/bootstrap.js']
      },
      sass: {
        src: ['<%= config.app %>/styles/{,*/}*.{scss,sass}'],
        ignorePath: /(\.\.\/){1,2}bower_components\//
      }
    },

    // Renames files for browser caching purposes
    rev: {
      dist: {
        files: {
          src: [
            '<%= config.dist %>/scripts/{,*/}*.js',
            '<%= config.dist %>/styles/{,*/}*.css',
            '<%= config.dist %>/images/{,*/}*.*',
            '<%= config.dist %>/styles/fonts/{,*/}*.*',
            '<%= config.dist %>/*.{ico,png}'
          ]
        }
      }
    },

    // Reads HTML for usemin blocks to enable smart builds that automatically
    // concat, minify and revision files. Creates configurations in memory so
    // additional tasks can operate on them
    useminPrepare: {
      options: {
        dest: '<%= config.dist %>'
      },
      html: '<%= config.app %>/index.html'
    },

    // Performs rewrites based on rev and the useminPrepare configuration
    usemin: {
      options: {
        assetsDirs: [
          '<%= config.dist %>',
          '<%= config.dist %>/images',
          '<%= config.dist %>/styles'
        ]
      },
      html: ['<%= config.dist %>/{,*/}*.html'],
      css: ['<%= config.dist %>/styles/{,*/}*.css']
    },

    // The following *-min tasks produce minified files in the dist folder
    imagemin: {
      dist: {
        files: [{
          expand: true,
          cwd: '<%= config.app %>/images',
          src: '{,*/}*.{gif,jpeg,jpg,png}',
          dest: '<%= config.dist %>/images'
        }]
      }
    },

    svgmin: {
      dist: {
        files: [{
          expand: true,
          cwd: '<%= config.app %>/images',
          src: '{,*/}*.svg',
          dest: '<%= config.dist %>/images'
        }]
      }
    },

    htmlmin: {
      dist: {
        options: {
          collapseBooleanAttributes: true,
          collapseWhitespace: true,
          conservativeCollapse: true,
          removeAttributeQuotes: true,
          removeCommentsFromCDATA: true,
          removeEmptyAttributes: true,
          removeOptionalTags: true,
          removeRedundantAttributes: true,
          useShortDoctype: true
        },
        files: [{
          expand: true,
          cwd: '<%= config.dist %>',
          src: '{,*/}*.html',
          dest: '<%= config.dist %>'
        }]
      }
    },

    // By default, your `index.html`'s <!-- Usemin block --> will take care
    // of minification. These next options are pre-configured if you do not
    // wish to use the Usemin blocks.
    // cssmin: {
    //   dist: {
    //     files: {
    //       '<%= config.dist %>/styles/main.css': [
    //         '.tmp/styles/{,*/}*.css',
    //         '<%= config.app %>/styles/{,*/}*.css'
    //       ]
    //     }
    //   }
    // },
    // uglify: {
    //   dist: {
    //     files: {
    //       '<%= config.dist %>/scripts/scripts.js': [
    //         '<%= config.dist %>/scripts/scripts.js'
    //       ]
    //     }
    //   }
    // },
    // concat: {
    //   dist: {}
    // },

    // Copies remaining files to places other tasks can use
    copy: {
      dist: {
        files: [{
          expand: true,
          dot: true,
          cwd: '<%= config.app %>',
          dest: '<%= config.dist %>',
          src: [
            '*.{ico,png,txt}',
            'images/{,*/}*.webp',
            '{,*/}*.html',
            'styles/fonts/{,*/}*.*'
          ]
        }, {
          src: 'node_modules/apache-server-configs/dist/.htaccess',
          dest: '<%= config.dist %>/.htaccess'
        }, {
          expand: true,
          dot: true,
          cwd: '.',
          src: 'bower_components/bootstrap-sass-official/assets/fonts/bootstrap/*',
          dest: '<%= config.dist %>'
        }]
      },
      styles: {
        expand: true,
        dot: true,
        cwd: '<%= config.app %>/styles',
        dest: '.tmp/styles/',
        src: '{,*/}*.css'
      }
    },

    // Generates a custom Modernizr build that includes only the tests you
    // reference in your app
    modernizr: {
      dist: {
        devFile: 'bower_components/modernizr/modernizr.js',
        outputFile: '<%= config.dist %>/scripts/vendor/modernizr.js',
        files: {
          src: [
            '<%= config.dist %>/scripts/{,*/}*.js',
            '<%= config.dist %>/styles/{,*/}*.css',
            '!<%= config.dist %>/scripts/vendor/*'
          ]
        },
        uglify: true
      }
    },

    // Run some tasks in parallel to speed up build process
    concurrent: {
      server: [
        'sass:server',
        'copy:styles'
      ],
      test: [
        'copy:styles'
      ],
      dist: [
        'sass',
        'copy:styles',
        'imagemin',
        'svgmin'
      ]
    }
    ,
    karma: {
      unit: {
        configFile: 'karma.conf.js',
        // run karma in the background
        background: process.env.TRAVIS ? true : false,
        // which browsers to run the tests on
        browsers: ['Chrome', 'Firefox']
      }
    }
  });

  grunt.registerTask('serve', 'start the server and preview your app, --allow-remote for remote access', function (target) {
    if (grunt.option('allow-remote')) {
      grunt.config.set('connect.options.hostname', '0.0.0.0');
    }
    if (target === 'dist') {
      return grunt.task.run(['build', 'connect:dist:keepalive']);
    }

    grunt.task.run([
      'clean:server',
      'wiredep',
      'concurrent:server',
      'autoprefixer',
      'connect:livereload',
      'watch'
    ]);
  });

  grunt.registerTask('server', function (target) {
    grunt.log.warn('The `server` task has been deprecated. Use `grunt serve` to start a server.');
    grunt.task.run([target ? ('serve:' + target) : 'serve']);
  });

  grunt.registerTask('build', [
    'clean:dist',
    'wiredep',
    'useminPrepare',
    'concurrent:dist',
    'autoprefixer',
    'concat',
    'cssmin',
    'uglify',
    'copy:dist',
    'modernizr',
    'rev',
    'usemin',
    'htmlmin'
  ]);

  grunt.registerTask('default', [
    'newer:jshint',
    'build'
  ]);

  // Travis CI task.
  //grunt.registerTask('travis', 'newer:jshint');
};
